src/Network/Transport/TCP.hs

```diff
@@ -999,33 +921,120 @@ handleConnectionRequest transport socketClosed (sock, sockAddr) = handle handleE
+      go ourEndPoint theirAddress = do
+        -- This runs in a thread that will never be killed
+        mEndPoint <- handle ((>> return Nothing) . handleException) $ do
+          resetIfBroken ourEndPoint theirAddress
+          (theirEndPoint, isNew) <-
+            findRemoteEndPoint ourEndPoint theirAddress RequestedByThem Nothing
+
+          if not isNew
+            then do
+              void $ tryIO $ sendMany sock
+                [encodeWord32 (encodeConnectionRequestResponse ConnectionRequestCrossed)]
+              probeIfValid theirEndPoint
+              return Nothing
+            else do
+              sendLock <- newMVar ()
+              let vst = ValidRemoteEndPointState
+                          {  remoteSocket        = sock
+                          ,  remoteSocketClosed  = socketClosed
+                          ,  remoteProbing       = Nothing
+                          ,  remoteSendLock      = sendLock
+                          , _remoteOutgoing      = 0
+                          , _remoteIncoming      = Set.empty
+                          , _remoteLastIncoming  = 0
+                          , _remoteNextConnOutId = firstNonReservedLightweightConnectionId
+                          }
+              sendMany sock [encodeWord32 (encodeConnectionRequestResponse ConnectionRequestAccepted)]
+              resolveInit (ourEndPoint, theirEndPoint) (RemoteEndPointValid vst)
+              return (Just theirEndPoint)
+        -- If we left the scope of the exception handler with a return value of
+        -- Nothing then the socket is already closed; otherwise, the socket has
+        -- been recorded as part of the remote endpoint. Either way, we no longer
+        -- have to worry about closing the socket on receiving an asynchronous
+        -- exception from this point forward.
+        forM_ mEndPoint $ handleIncomingMessages (transportParams transport) . (,) ourEndPoint
+
+      probeIfValid :: RemoteEndPoint -> IO ()
+      probeIfValid theirEndPoint = modifyMVar_ (remoteState theirEndPoint) $
+        \st -> case st of
+          RemoteEndPointValid
+            vst@(ValidRemoteEndPointState { remoteProbing = Nothing }) -> do
+              tid <- forkIO $ do
+                -- send probe
+                let params = transportParams transport
+                void $ tryIO $ System.Timeout.timeout
+                    (maybe (-1) id $ transportConnectTimeout params) $ do
+                  sendMany (remoteSocket vst)
+                    [encodeWord32 (encodeControlHeader ProbeSocket)]
+                  threadDelay maxBound
+                -- Discard the connection if this thread is not killed (i.e. the
+                -- probe ack does not arrive on time).
+                --
+                -- The thread handling incoming messages will detect the socket is
+                -- closed and will report the failure upwards.
+                tryCloseSocket (remoteSocket vst)
```

Comment:
In this case the connecting peer tries to make a new socket when there already is one. The new socket is closed up (`handleConnectionRequest` returns after sending the connection crossed header) and then a probe is sent on the existing socket `remoteSocket vst`.

> I don't see how the connection is discarded if TCP_USER_TIMEOUT is not set. So perhaps there is an unhandled case here.

That's done here: if there's no `ProbeSocketAck` before the timeout, the socket is closed and the `handleIncomingMessages` thread will eventually stop with `prematureExit`.

> I don't think it can be restored without reintroducing the race that patch fixes.

I think it's OK because the `MVar` created by `forkServer` when a connection is accepted won't be filled until the continuation returns. Closing the socket here doesn't fill it, it just causes `handleIncomingMessages` to stop which in turn causes the `MVar` to be filled.